File: tree/104-maximum-depth-of-binary-tree.cpp
// Solution 1: recursion
// N = number of nodes in tree
// O(N) in time
// O(N) in space in worst case (skewed tree)
// O(N) in space in average
// pattern: pre-order
class Solution {
public:
    int maxDepth(TreeNode* root) {
        if (root == NULL)
            return 0;
        return max(maxDepth(root->left), maxDepth(root->right)) + 1;
    }
};

// Solution 2: iterative using BFS
// Time & Space complexity same as solution 1
class Solution {
public:
    int maxDepth(TreeNode* root) {
        // Base case
        if (root == NULL) return 0;

        int depth = 0;
        stack<TreeNode*> currLevel;
        currLevel.push(root);
        while(!currLevel.empty()) {
            stack<TreeNode*> nextLevel;
            depth++;

            // Traverse all nodes in current level
            while(!currLevel.empty()) {
                TreeNode* node = currLevel.top();
                currLevel.pop();

                // Save nodes in the next level
                if (node->left != NULL) nextLevel.push(node->left);
                if (node->right != NULL) nextLevel.push(node->right);
            }

            // Bfs to the next level
            currLevel = nextLevel;
        }

        return depth;
    }
};

File: practices/leetcode/tree/104-maximum-depth-of-binary-tree.py
# Solution 1: same as solution 1 in C++ version
class Solution:
    def maxDepth(self, root):
        if root is None:
            return 0
        return max(self.maxDepth(root.left), self.maxDepth(root.right)) + 1

# Solution 2: same as solution 2 in C++ version
class Solution:
    def maxDepth(self, root):
        # Base case
        if root is None:
            return 0
        levels = [root]
        depth = 0
        while levels:
            depth += 1
            levels = [kid for node in levels for kid in (node.left, node.right) if kid]
        return depth
        
